File: src/main/scala/uk/co/ridentbyte/view/CardinalView.scala
package uk.co.ridentbyte.view

import java.net.{ConnectException, URISyntaxException, UnknownHostException}
import java.util.function

import javafx.application.Platform
import javafx.scene.control.SplitPane
import javafx.scene.layout.{BorderPane, GridPane, Priority}
import javax.net.ssl.SSLHandshakeException
import uk.co.ridentbyte.model.{CardinalRequest, CardinalResponse, Config}
import uk.co.ridentbyte.view.request.{RequestControlPane, RequestInputPane}
import uk.co.ridentbyte.view.response.ResponsePane
import uk.co.ridentbyte.view.util.{ColumnConstraintsBuilder, RowConstraintsBuilder}

import scala.runtime.BoxedUnit

class CardinalView(showAsCurl: java.util.function.Function[Void, Void],
                   showErrorDialogCallback: String => BoxedUnit,
                   getConfigCallback: () => Config,
                   exportToCsv: List[(CardinalRequest, Option[CardinalResponse])] => Unit,
                   exportToBash: (List[CardinalRequest], Option[Long]) => Unit,
                   sendRequestCallback: CardinalRequest => CardinalResponse,
                   triggerUnsavedChangesMade: java.util.function.Function[Void, Void]) extends BorderPane {

  private val requestResponseSplitPane = new SplitPane()
  requestResponseSplitPane.setDividerPositions(0.4)

  private val requestInputPane = new RequestInputPane(triggerUnsavedChangesMade)
  requestInputPane.setMinWidth(400)
  private val responsePane = new ResponsePane(getConfigCallback, sendRequestCallback, exportToCsv, exportToBash, showErrorDialogCallback)
  responsePane.setMinWidth(400)
  private val requestControlPane = new RequestControlPane(showAsCurl, showBulkRequestInput, sendSingleRequest)

  val grid = new GridPane
  grid.getColumnConstraints.add(new ColumnConstraintsBuilder().withHgrow(Priority.ALWAYS).build)
  grid.getRowConstraints.add(new RowConstraintsBuilder().withVgrow(Priority.ALWAYS).build)
  grid.getRowConstraints.add(new RowConstraintsBuilder().withVgrow(Priority.NEVER).build)

  requestResponseSplitPane.getItems.add(requestInputPane)
  requestResponseSplitPane.getItems.add(responsePane)

  grid.add(requestResponseSplitPane, 0, 0)

//  GridPane.setColumnSpan(requestControlPane, 2)
  grid.add(requestControlPane, 0, 1)

  setCenter(grid)

  def clearAll(): Unit = {
    requestInputPane.clear()
    responsePane.clearContents()
  }

  private def sendSingleRequest(onStart: () => BoxedUnit, onFinish: () => BoxedUnit): BoxedUnit = {
    Platform.runLater(() => responsePane.clearContents())
    val request = requestInputPane.getRequest
    if (request.uri.trim.length == 0) {
      showErrorDialogCallback("Please enter a URL.")
    } else {
      new Thread() {
        override def run(): Unit = {
          Platform.runLater(() => onStart())
          val httpResponse = try {
            Some(sendRequestCallback(requestInputPane.getRequest.processConstants(getConfigCallback())))
          } catch {
            case _: ConnectException => showErrorDialogCallback("Connection refused."); None
            case _: URISyntaxException => showErrorDialogCallback("Invalid URL."); None
            case _: UnknownHostException => showErrorDialogCallback("Unknown Host."); None
            case _: SSLHandshakeException => showErrorDialogCallback("SSL Handshake failed. Remote host closed connection during handshake."); None
            case e: Exception =>
              e.printStackTrace()
              showErrorDialogCallback("Unknown error occurred."); None
          } finally {
            Platform.runLater(() => { onFinish() })
          }
          responsePane.setResponse(httpResponse)
        }
      }.start()
      scala.runtime.BoxedUnit.UNIT
    }
  }

  private def showBulkRequestInput: java.util.function.Function[Void, Void] = {
    new function.Function[Void, Void] {
      override def apply(t: Void): Void = {
        val request = requestInputPane.getRequest
        if (request.uri.trim.length == 0) {
          showErrorDialogCallback("Please enter a URL.")
          null
        } else {
          responsePane.showBulkRequestInput(requestInputPane.getRequest)
          scala.runtime.BoxedUnit.UNIT
          null
        }
      }
    }
  }

  def loadRequest(request: CardinalRequest): Unit = requestInputPane.loadRequest(request)

  def loadCurlCommand(curl: String): Unit = responsePane.loadCurlCommand(curl)

  def getRequest: CardinalRequest = requestInputPane.getRequest

  def setBody(body: String): Unit = requestInputPane.setBody(body)

  def addHeader(header: String): Unit = requestInputPane.addHeader(header)

}

File: src/main/java/uk/co/ridentbyte/view/CardinalView.java
package uk.co.ridentbyte.view;

import javafx.application.Platform;
import javafx.scene.control.SplitPane;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.Priority;
import uk.co.ridentbyte.functions.ExportToBash;
import uk.co.ridentbyte.functions.ShowAsCurl;
import uk.co.ridentbyte.functions.UnsavedChangesMade;
import uk.co.ridentbyte.model.CardinalRequestAndResponse;
import uk.co.ridentbyte.model.CardinalRequest;
import uk.co.ridentbyte.model.CardinalResponse;
import uk.co.ridentbyte.model.Config;
import uk.co.ridentbyte.model.Vocabulary;
import uk.co.ridentbyte.view.request.RequestControlPane;
import uk.co.ridentbyte.view.request.RequestInputPane;
import uk.co.ridentbyte.view.response.ResponsePane;
import uk.co.ridentbyte.view.util.ColumnConstraintsBuilder;
import uk.co.ridentbyte.view.util.RowConstraintsBuilder;

import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Function;

public class CardinalView extends BorderPane {

    private RequestInputPane requestPane;
    private ResponsePane responsePane;

    private Function<String, Void> showErrorDialog;
    private Function<CardinalRequest, CardinalResponse> sendRequest;
    private Function<Void, Config> getConfig;
    private final RequestControlPane requestControlPane;

    public CardinalView(ShowAsCurl showAsCurl,
                        Function<String, Void> showErrorDialog,
                        Function<Void, Config> getConfig,
                        Function<List<CardinalRequestAndResponse>, Void> exportToCsv,
                        ExportToBash exportToBash,
                        Function<CardinalRequest, CardinalResponse> sendRequest,
                        UnsavedChangesMade triggerUnsavedChangesMade,
                        Vocabulary vocabulary) {

        this.showErrorDialog = showErrorDialog;
        this.sendRequest = sendRequest;
        this.getConfig = getConfig;

        SplitPane requestResponseSplitPane = new SplitPane();
        requestResponseSplitPane.setDividerPositions(0.4);

        this.requestPane = new RequestInputPane(triggerUnsavedChangesMade, vocabulary);
        requestPane.setMinWidth(400);

        this.responsePane = new ResponsePane(getConfig, sendRequest, exportToCsv, exportToBash, showErrorDialog);
        responsePane.setMinWidth(400);

        requestControlPane = new RequestControlPane(showAsCurl, this.showBulkRequestInput(), this.sendSingleRequest());

        GridPane grid = new GridPane();
        grid.getColumnConstraints().add(new ColumnConstraintsBuilder().withHgrow(Priority.ALWAYS).build());
        grid.getRowConstraints().add(new RowConstraintsBuilder().withVgrow(Priority.ALWAYS).build());
        grid.getRowConstraints().add(new RowConstraintsBuilder().withVgrow(Priority.NEVER).build());

        requestResponseSplitPane.getItems().add(requestPane);
        requestResponseSplitPane.getItems().add(responsePane);

        grid.add(requestResponseSplitPane, 0, 0);
        grid.add(requestControlPane, 0, 1);

        this.setCenter(grid);
    }

    public void clearAll() {
        this.requestPane.clear();
        this.responsePane.clear();
    }

    public void loadRequest(CardinalRequest request) {
        this.requestPane.loadRequest(request);
    }

    public void loadCurlCommand(String curl) {
        this.responsePane.loadCurlCommand(curl);
    }

    public CardinalRequest getRequest() {
        return this.requestPane.getRequest();
    }

    public void setBody(String body) {
        this.requestPane.setBody(body);
    }

    public void addHeader(String header) {
        this.requestPane.addHeader(header);
    }

    private Function<Void, Void> showBulkRequestInput() {
        return new Function<Void, Void>() {
            @Override
            public Void apply(Void aVoid) {
                CardinalRequest request = requestPane.getRequest();
                if (request.getUri().trim().length() == 0) {
                    showErrorDialog.apply("Please enter a URL.");
                } else {
                    responsePane.showBulkRequestInput(requestPane.getRequest());
                }
                return null;
            }
        };
    }

    private BiFunction<Function<Void, Void>, Function<Void, Void>, Void> sendSingleRequest() {
        return new BiFunction<Function<Void, Void>, Function<Void, Void>, Void>() {
            @Override
            public Void apply(Function<Void, Void> onStart, Function<Void, Void> onFinish) {
                Platform.runLater(() -> responsePane.clear());
                CardinalRequest request = requestPane.getRequest();
                if (request.getUri().trim().length() == 0) {
                    showErrorDialog.apply("Please enter a URL.");
                } else {
                    new Thread(() -> {
                        Platform.runLater(() -> onStart.apply(null));
                        try {
                            CardinalResponse response = sendRequest.apply(requestPane.getRequest().processConstants(getConfig.apply(null)));
                            responsePane.setResponse(response);
                        } catch (Exception e) {
                            e.printStackTrace();
                            showErrorDialog.apply("Unknown error");
                        } finally {
                            Platform.runLater(() -> onFinish.apply(null));
                        }
                    }).start();
                }
                return null;
            }
        };
    }

    public void triggerSendRequest() {
        requestControlPane.triggerSendRequest();
    }

}
